manager/supermanager.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2021 VMware, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package manager
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// New returns a controller manager which can subquently create other managers
+// with the given REST configuration and controller options. If the REST
+// configuration or anything it refers to is mutated or if anything the
+// controller options refers to is subsequently mutated, unpredictable behaviour
+// may occur.
+func New(config *rest.Config, options ctrl.Options) (SuperManager, error) {
+	mainManager, err := ctrl.NewManager(config, options)
+	if err != nil {
+		return nil, err
+	}
+
+	return &superManager{
+		Manager:     mainManager,
+		config:      config,
+		options:     options,
+		failures:    make(chan error),
+		nextContext: big.NewInt(0),
+		contexts:    make(map[*big.Int]context.CancelFunc),
+	}, nil
+}
+
+// SuperManager is a controller manager which can create other managers in order
+// to be able to stop watches by stopping the corresponding caches.
+type SuperManager interface {
+	ctrl.Manager
+
+	// NewManager returns a new controller manager which is configured the same
+	// as this super manager.
+	NewManager() (ctrl.Manager, error)
+
+	// StartController starts the given controller asynchronously and returns a
+	// cancel function which may be used to stop the controller.
+	StartController(ctrl controller.Controller) context.CancelFunc
```

Comment:
calls to a "Start" method typically block until canceled or errored. What this method is doing is more equivalent to `Add` on the manager. Perhaps something more like AddCancelable to indicate there is a new runnable with a different lifecycle.

```suggestion
	AddCancelable(r manager.Runnable) (context.CancelFunc, error)
```

or, we can pass in a cancelable context rather than creating it within the method.

```suggestion
	AddWithContext(ctx context.Context, r manager.Runnable) error
```

Reply:
Good idea. I went with `AddCancelable` so the supermanager can create an appropriate `cancelFunc` without needing to provide another method to do this.